Source/MQTTnet.Extensions.ManagedClient/ManagedMqttClient.cs

```diff
@@ -408,6 +408,7 @@ private void TryPublishQueuedMessage(ManagedMqttApplicationMessage message)
                     {
                         _messageQueue.RemoveFirst(i => i.Id.Equals(message.Id));
                     }
+                    _storageManager?.RemoveAsync(message).GetAwaiter().GetResult();
```

Comment:
This will throw a null ref exception in case of a StorageManager is missing because the ? operator will return null and then "GetAwaiter" is called on _null_.

Also I would prefer to change the whole method chain to _async_ instead of using _GetResult()_.

And I would like to have a UnitTest for the future to ensure that this is fixed.

What do you think? Do you want to implement that as well?

Reply:
@chkr1011 From https://msdn.microsoft.com/en-us/magazine/dn802602.aspx

> The null-conditional operator conditionally checks for null before invoking the target method and any additional method within the call chain.

See https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEUCuA7NAExAGoAfAAQCYBGAWAChqAGR6gZgAIBvRz/ztSo8+AgJABlDFACWeAOacAzpwC8nPDgA2WgNyiBAijQCcACgBEgHg3AIPshOFziU5mzSgPwA6CTmDHmZlQAlN6+/oFBQZzu7g4AcgCqADJJFpH6DIYAvoxZQA

Yes a test is always a good idea, it could have prooved the fix to work ;)

Reply:
Oh yes you are right. I was thinking of await which will throw an exception if the task is null (which may happen when using the ? operator).

Reply:
Yes this is an issue. But I guess we have another issue here. If no storage manager is assigned this current code will never succeed and always throw an exception in line 390.
https://github.com/chkr1011/MQTTnet/blob/f76e13028316e3d1757d871bb9f200945a72583d/Source/MQTTnet.Extensions.ManagedClient/ManagedMqttClient.cs#L379-L390
I copied this line from 390 to line 411
https://github.com/chkr1011/MQTTnet/blob/f76e13028316e3d1757d871bb9f200945a72583d/Source/MQTTnet.Extensions.ManagedClient/ManagedMqttClient.cs#L398-L412
I'll try to fix this and provide a test on Monday.

Reply:
OK, then let me merge it yet. I will have a look today because I have some free time and want to release 3.0.3 soon. Then this change will be already part of it.